text/0010-OIDC.md

```diff
@@ -0,0 +1,230 @@
+
+<details>
+  <summary>Detailed implementation notes</summary>
+
+- new provider table, that contains the stuff in https://token.actions.githubusercontent.com/.well-known/openid-configuration & https://token.actions.githubusercontent.com/.well-known/jwks
+- new table that’s something like a _role_. it binds a provider & user & token created policy
+- a JWT token table, where we can store the details of the JWT from the OIDC issuer and link it to the APIKey it creates
+- and the policy would probably be { scopes: [], rubygems: [], conditions: [] }
+- and then, to get the token, there’d be an API endpoint _for the role_ that takes a POST with the JWT, looks up the role, decodes & validates the JWT, and then checks the claims in the JWT against the conditions, and if it all matches, return an API key
+- and we probably want to add an expires_at to the API key table as well
+- oh and we probably also want to track the API key used to push a version, and then only ever soft-delete the API keys
+- that way, if its pushed using OIDC, we can (down the road) expose some of the claims
+
+</details>
+
+# Drawbacks
+
+This is a large change to rubygems.org!
+
+In particular, it puts us down the path of "soft deleting" records such as API tokens, instead of removing them from the DB entirely.
+That's a paradigm shift (though one we probably wanted anyway), and it requires updating the way we interact with those records
+to ensure we only allow using "undeleted" rows.
+
+# Rationale and Alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+The impact of not adding OIDC support to RubyGems.org would be limiting the platform's ability to integrate with modern authentication systems and identity providers. Users who rely on OIDC for authentication would not be able to easily and securely interact with RubyGems.org using their existing identity provider. This could lead to user frustration, reduced adoption of the platform, and potential security risks if users resort to less secure authentication mechanisms as workarounds. By implementing OIDC support, RubyGems.org enhances its usability, security, and interoperability, enabling seamless integration with a wide range of identity providers and aligning with industry best practices.
+
+# Unresolved questions
+
+- Is the schema for Access Policies extensible enough to support arbitrary OIDC providers / the claims they will sign?
+- What does the design for this look like when we allow onboarding Providers / API Key Roles through the web UI?
+- How should we protect a future API that allows for creating & updating API key roles?
+- What claims, if any, should we surface in the UI / API? What should that look like?
```

Comment:
References:
* Related discussion in PyPI - https://github.com/pypi/warehouse/issues/14799
* Claims available from GitHub Actions in particular: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token

---

I'm most interested in the `repository ` and `job_workflow_ref` claim, like `octo-org/octo-automation/.github/workflows/oidc.yml@refs/heads/main`.

I want that so:
* I can compare to the repository specified by the gem.
* I can compare the `refs/tags/v1.2.3` portion to the version of the gem that was generated. That makes me comfortable the gem wasn't published by an `issue_comment` event that allowed script injection.
* I can fetch and analyze the workflow file that produced the release
    * Maybe I'll make sure there is only one `.jobs[]` with `.permissions["id-token"]: write`
    * Maybe I'll look at the `.job[].steps[].uses` and match against an allowlist (e.g. I want the workflow to contain only `actions/checkout@.*`, `ruby/setup-ruby@.*` and `rubygems/release-gem@.*` steps.
    * The code in https://github.com/ossf/scorecard-webapp/blob/main/app/server/verify_workflow.go is a good implementation of ways one might want to verify the workflow file doesn't contain any shenaningans.
* I especially like how this attribute works with [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow).  If https://github.com/rubygems/release-gem included a reusable workflow for users instead of an example workflow to copy/paste, then I could match against `job_workflow_ref: rubygems/release-gem/.github/workflows/release.yml@.*` to allow any gems published using the "approved" workflow.

Please redirect if there is a better place for this discussion ❤️ .